std-lib/list/list.lurk

```diff
@@ -0,0 +1,62 @@
+;; a list is of the form (cons a (cons b (cons ... (cons nil)...)))
+;; some definitions for lists 
+
+;; empty_list : the empty list
+(let ((empty_list '())) (current-env))
+
+;; some functions for lists 
+
+;; list : make a list out of a (possibly unevaluated) list quote
+(letrec ((list (lambda (q)
+		 (if (car q)
+		     (cons (eval (car q)) (list (cdr q)))
+		     '()))))
+  (current-env))
+
+;; list_eq : determine if two lists are equal
+(letrec ((list_eq (lambda (l1 l2)
+    (if (car l1) 
+        (if (car l2)
+            (if (= (car l1) (car l2))
+                (list_eq (cdr l1) (cdr l2))
+                nil)
+            nil)
+        (if (car l2)
+            nil
+            t)))))
+    (current-env))
+
+;; snoc : appends to the right-hand side of a list (opposite of cons)
+(letrec 
+    ((snoc_iter (lambda (old_list new_list) 
+        (if (car old_list)
+            (snoc_iter (cdr old_list) (cons (car old_list) new_list))
+            new_list)))
+    (snoc (lambda (x lst) (snoc_iter (snoc_iter lst '()) (cons x nil))))) 
+    (current-env))
+
+;; list_length : determines the length of a list 
+(letrec 
+    ((length_iter (lambda (lst cntr)
+        (if (car lst)
+            (length_iter (cdr lst) (+ 1 cntr))
+            cntr)))
+    (list_length (lambda (lst) (length_iter lst 0))))
+    (current-env))
+
+;; list_map : a list map function 
+(letrec 
+    ((map_iter (lambda (lst fun accum)
+        (if (car lst)
+            (map_iter (cdr lst) fun (snoc (fun (car lst)) accum))
+            accum)))
+    (list_map (lambda (lst fun) (map_iter lst fun (cons nil nil)))))
+    (current-env))
+
+;; list_fold : a fold function over lists 
+(letrec 
+    ((list_fold (lambda (lst op accum)
+        (if (car lst)
+            (list_fold (cdr lst) op (op (car lst) accum))
+            accum))))
+    (current-env))
```

Comment:
Same here - suggest the `op` argument goes first so you can do this:

```
  (letrec 
      ((list_fold (lambda (op accum lst)
          (if (car lst)
              (list_fold op (op (car lst) accum) (cdr lst))
              accum)))
       (sum (list_fold (lambda (x y) (+ x y)) 0)))
    (sum '(1 2 3 4 5)))
=> 15
```

(by the way I notice `+` isn't a first class function, so I had to introduce the lambda there, not sure if this is something I should open as a separate lurk issue or if it's intentional.)

Reply:
No, it's how it works. All the built-in operators are hardcoded in the circuit and are not actually 'functions'. We could define functions bound to the symbols in a library in the future, as a convenience — or not. Depending on how details of future versions evolve, along with the circuit, it's *possible* we revisit this, but for now it's how things are by design.

Reply:
Ah good to know, I had the same question about `+`